### PR TITLE
Soft task

### PR DESCRIPTION
**Pull request by TheSpaceSheep**
- implement the possibility to choose the number of masked attributes within a batch randomly (according to a predefined distribution).
- make plateau threshold for early stopping a command line argument

### CODE REVIEW

File: egg/zoo/compo_vs_generalization/curriculum_games.py [hunk 1]
```diff
@@ -62,43 +63,71 @@ def __init__(self,
...
+        if self.reveal_distribution == 'deterministic':
+            probs = torch.zeros((batch_size, self.curriculum_level))
+            probs[:,  -1] = 1.
+        elif self.reveal_distribution == 'uniform':
+            probs = torch.ones((batch_size, self.curriculum_level)) / self.curriculum_level
+        else:
+            raise NotImplemented
+
+        distr = torch.distributions.Categorical(probs)
+        n_revealed = distr.sample()
```
Comment (pmichel31415) [inline egg/zoo/compo_vs_generalization/curriculum_games.py]: No need for torch.distributions, you can use 

```python
n_revealed = torch.multinomial(probs, 1)
```

File: egg/zoo/compo_vs_generalization/curriculum_games.py [hunk 2]
```diff
@@ -62,43 +63,71 @@ def __init__(self,
...
+        n_revealed += 1  # to prevent 0 revealed
```
Comment (pmichel31415) [inline egg/zoo/compo_vs_generalization/curriculum_games.py]: Why?

File: egg/zoo/compo_vs_generalization/curriculum_games.py [hunk 3]
```diff
@@ -62,43 +63,71 @@ def __init__(self,
...
-            else:
-                idxs_to_mask = torch.multinomial(mask_probability,
-                        n_masks,
-                        replacement=False)
-
-        idxs_to_mask = idxs_to_mask.to(sender_input.device)
+            reveal_probability = torch.ones((batch_size, self.n_attributes))/self.n_attributes
+            idxs_to_reveal = torch.multinomial(reveal_probability,
+                    self.n_attributes,
+                    replacement=False)
```
Comment (pmichel31415) [inline egg/zoo/compo_vs_generalization/curriculum_games.py:103]: Nice way to sample permutations :+1:

File: egg/zoo/compo_vs_generalization/curriculum_games.py [hunk 4]
```diff
@@ -62,43 +63,71 @@ def __init__(self,
...
+
+        # mask indices with a redundant value
+        mask_idxs_to_reveal = torch.arange(self.n_attributes).expand(
+            batch_size, self.n_attribute
+        )
```
Comment (pmichel31415) [inline egg/zoo/compo_vs_generalization/curriculum_games.py]: Doesn't this raise an exception? You are expanding a non-existing dimension. I thought you would need to call `.view(1, self.n_attributes)` to create a singleton dimension to expand.

File: egg/zoo/compo_vs_generalization/curriculum_games.py (hunk 2 above)
Comment (pmichel31415) [reply egg/zoo/compo_vs_generalization/curriculum_games.py]: I think your comment confused me. I guess you are adding `+1` because you want `n_revealed` to be between 1 and n_attributes, right?

File: egg/zoo/compo_vs_generalization/curriculum_games.py [hunk 5]
```diff
@@ -62,43 +63,71 @@ def __init__(self,
...
+        mask_idxs_to_reveal = (
+                mask_idxs_to_reveal < n_revealed.view(batch_size, -1)
+        ).long()
+        idxs_to_reveal *= mask_idxs_to_reveal
+        idxs_to_reveal += idxs_to_reveal[:, 0].view(
+            batch_size, 1
+        ).expand(batch_size, self.n_attributes) * (1 - mask_idxs_to_reveal)
+        
+        idxs_to_reveal = idxs_to_reveal.to(sender_input.device)
+        n_revealed = n_revealed.to(sender_input.device)
```
Comment (pmichel31415) [inline egg/zoo/compo_vs_generalization/curriculum_games.py]: You should probably place everything on the device at the beginning (even before you do all the masking manipulations)

File: egg/zoo/compo_vs_generalization/curriculum_games.py [hunk 6]
```diff
@@ -62,43 +63,71 @@ def __init__(self,
...
+        mask_idxs_to_reveal = torch.arange(self.n_attributes).expand(
+            batch_size, self.n_attribute
+        )
+        mask_idxs_to_reveal = (
+                mask_idxs_to_reveal < n_revealed.view(batch_size, -1)
+        ).long()
+        idxs_to_reveal *= mask_idxs_to_reveal
+        idxs_to_reveal += idxs_to_reveal[:, 0].view(
+            batch_size, 1
+        ).expand(batch_size, self.n_attributes) * (1 - mask_idxs_to_reveal)
```
Comment (pmichel31415) [inline egg/zoo/compo_vs_generalization/curriculum_games.py]: Might be worth adding a couple more comments here to explain what's happening. That last line in particular is really not clear

File: egg/zoo/compo_vs_generalization/curriculum_games.py [hunk 7]
```diff
@@ -62,43 +63,71 @@ def __init__(self,
...
+
+        # mask indices with a redundant value
+        mask_idxs_to_reveal = torch.arange(self.n_attributes).expand(
+            batch_size, self.n_attribute
+        )
+        mask_idxs_to_reveal = (
+                mask_idxs_to_reveal < n_revealed.view(batch_size, -1)
+        ).long()
+        idxs_to_reveal *= mask_idxs_to_reveal
+        idxs_to_reveal += idxs_to_reveal[:, 0].view(
```
Comment (pmichel31415) [inline egg/zoo/compo_vs_generalization/curriculum_games.py]: I suggest taking the good habit of avoiding in-place operations as much as possible in pytorch. It doesn't make a difference here but it can really trip you up if you're not careful

File: egg/zoo/compo_vs_generalization/losses.py [hunk 8]
```diff
@@ -120,58 +121,79 @@ def forward(
...
+                    (
+                        # this counts all masks as correct predictions
+                        masked_output.argmax(dim=-1) == masked_input.argmax(dim=-1)
+                    ).detach(),
+                    dim=1,
+                )
+                # so we have to compensate by
+                # substracting the number of masks
+                - (self.n_attributes - n_revealed)
             )
```
Comment (pmichel31415) [inline egg/zoo/compo_vs_generalization/losses.py]: This is very hard to read. There is also a lot of redundancy because you are
1. masking the inputs and predictions
2. Computing the losses/accuracies, etc based on these masked input/output
3. Handling the masked positions **again** in the losses/accuracies

Why not skip step 1 since you need to handle the masked attributes at the end anyway? Compute loss/acc for every attribute, and only apply the masks when you are aggregating the metric. This would save you a lot of lines of code.

How about this:

```python
receiver_input = receiver_input.view(batch_size, self.n_attributes, self.n_values)
receiver_output = receiver_output.view(batch_size, self.n_attributes, self.n_values)
# Compute mask for revealed attributes
revealed_mask = torch.zeros_like(receiver_input)
revealed_mask = revealed_mask.scatter(dim=1, index=idxs_to_reveal, value=1).float()
# Matches for each attributes
matches = (receiver_output.argmax(dim=-1) == receiver_input.argmax(dim=-1)).float()
# Average across attributes (but only count revealed attributes)
acc_or = (matches * revealed_mask).sum(-1) / revealed_mask.sum(-1)
# Exact matches: Count 1 if attribute is predicted correctly OR if it is masked.
acc = torch.all((matches == 1) | (revealed_mask == 0), dim=-1).float()
# Loss for each attribute (you need to flatten the first dimensions and reshape them afterwards)
loss_by_attributes = F.cross_entropy(
    receiver_output.view(-1, n_attributes),
    receiver_input.argmax(-1).view(-1),
    reduction="none",
).view(batch_size, n_attributes)
# Take the mean, but only of the revealed attributes
loss = (losses * revealed_mask).sum(-1) / revealed_mask.sum(-1)
```

**Review by pmichel31415 — CHANGES_REQUESTED** [review]
Overall looks good. However, there is a lot of complexity starting to pile up in the function where you compute the loss. I've proposed a an alternative that should do the same thing but in ~10 lines of code. Please refer to my comment there directly.

File: egg/zoo/compo_vs_generalization/curriculum_games.py [hunk 9]
```diff
@@ -102,21 +102,26 @@ def forward(self, sender_input, labels, receiver_input=None, aux_input=None):
...
-        )
+            batch_size, self.n_attributes
+        ).to(device)
         mask_idxs_to_reveal = (
                 mask_idxs_to_reveal < n_revealed.view(batch_size, -1)
         ).long()
-        idxs_to_reveal *= mask_idxs_to_reveal
-        idxs_to_reveal += idxs_to_reveal[:, 0].view(
+        idxs_to_reveal = idxs_to_reveal * mask_idxs_to_reveal
+        idxs_to_reveal = idxs_to_reveal + idxs_to_reveal[:, 0].view(
```
Comment (pmichel31415) [inline egg/zoo/compo_vs_generalization/curriculum_games.py:121]: Maybe do everything with a batch_size x n_attributes mask from here onwards?
